src/coreclr/jit/importer.cpp

```diff
@@ -16529,19 +16544,40 @@ void Compiler::impImportBlockCode(BasicBlock* block)
                     break;
                 }
 
+                bool isByRefLike =
+                    (info.compCompHnd->getClassAttribs(resolvedToken.hClass) & CORINFO_FLG_BYREF_LIKE) != 0;
+                if (isByRefLike)
+                {
+                    // For ByRefLike types we are required to either fold the
+                    // recognized patterns in impBoxPatternMatch or otherwise
+                    // throw InvalidProgramException at runtime. In either case
+                    // we will need to spill side effects of the expression.
+                    impSpillSideEffects(false, CHECK_SPILL_ALL DEBUGARG("Required for box of ByRefLike type"));
+                }
+
                 // Look ahead for box idioms
-                int matched = impBoxPatternMatch(&resolvedToken, codeAddr + sz, codeEndp);
+                int matched =
+                    impBoxPatternMatch(&resolvedToken, codeAddr + sz, codeEndp, isByRefLike ? IsByRefLike : None);
                 if (matched >= 0)
                 {
                     // Skip the matched IL instructions
                     sz += matched;
                     break;
                 }
 
-                impImportAndPushBox(&resolvedToken);
-                if (compDonotInline())
+                if (isByRefLike)
                 {
-                    return;
+                    // ByRefLike types are supported in boxing scenarios when the instruction can be elided
+                    // due to a recognized pattern above. If the pattern is not recognized, the code is invalid.
+                    BADCODE("ByRefLike types cannot be boxed");
```

Comment:
This will throw at JIT time even if the box cannot be reached at runtime. Is that intended?

Reply:
Yep. Part of the offline conversation we had was that all we really need to do is ensure the patterns match when a ByRefLike type is involved. Thanks for confirming.